src/data/quaternion.lean

```diff
@@ -0,0 +1,528 @@
+/-
+Copyright (c) 2020 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+-/
+
+import tactic.ring_exp ring_theory.algebra algebra.opposites algebra.commute data.equiv.ring
+
+/-!
+# Quaternions
+
+In this file we define quaternions `ℍ[R]` over a commutative ring `R`, and define some
+algebraic structures on `ℍ[R]`.
+
+## Main definitions
+
+* `quaternion_algebra R a b` :
+  [quaternion algebra](https://en.wikipedia.org/wiki/Quaternion_algebra) with coefficients `a`, `b`
+* `quaternion R`, `ℍ[R]` : the space of quaternions, a.k.a. `quaternion_algebra R (-1) (-1)`;
+* `quaternion.norm_sq` : square of the norm of a quaternion;
+* `quaternion.conj` : conjugate of a quaternion;
+
+We also define the following algebraic structures on `ℍ[R]`:
+
+* `ring ℍ[R, a, b]` and `algebra R ℍ[R, a, b]` : for any commutative ring `R`;
+* `ring ℍ[R]` and `algebra R ℍ[R]` : for any commutative ring `R`;
+* `domain ℍ[R]` : for a linear ordered commutative ring `R`;
+* `division_algebra ℍ[R]` : for a linear ordered field `R`.
+
+## Notation
+
+* `ℍ[R]` : quaternions over `R`.
+
+## Implementation notes
+
+We define quaternions over any ring `R`, not just `ℝ` to be able to deal with, e.g., integer
+or rational quaternions without using real numbers. In particular, all definitions in this file
+are computable.
```

Comment:
So, shall we do this for the complex numbers too?

Reply:
We already have zsqrtd. We could merge it with `complex` possibly generalizing to a "nice two-fields structure for the quotient ring `R[x] / (x^2+ax+b)`".

Reply:
Sounds good, for at some distant point in the future. :-)